### ISSUE TITLE
Allow JSON payload for GET calls

### ISSUE THREAD

**Issue by stevesuh** [tiangolo/fastapi#2004]
### First check

* [x ] I added a very descriptive title to this issue.
* [x ] I used the GitHub search to find a similar issue and didn't find it.
* [ x] I searched the FastAPI documentation, with the integrated search.
* [ x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x ] I already read and followed all the tutorial in the docs and didn't find an answer.
* [ x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [ x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [ x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x ] After submitting this, I commit to:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * Or, I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
from fastapi import FastAPI

app = FastAPI()


class Sample(BaseModel):
    countryId: int

@router.get("/foo/bar",
            response_model=Sample)
def get_sample(sample: Sample):
    return {"Hello": "World"}

### Description

* Open the browser and call the GET endpoint `/foo/bar`.
* Pass in JSON body with `{"countryId": 1}`.
* TypeError: Failed to execute 'fetch' on 'Window': Request with GET/HEAD method cannot have body.

### The solution you would like

It doesn't throw the TypeError and allows GET calls to pass in JSON payloads

### Environment

* OS: [e.g. Linux / Windows / macOS]: macOS
* FastAPI Version [e.g. 0.3.0]: 0.60.1


* Python version:

3.7.7

### Additional context

![Screen Shot 2020-09-02 at 2 00 10 PM](https://user-images.githubusercontent.com/5025518/92019556-ca7cbf00-ed24-11ea-84d6-52dbc18a8aec.png)

**Comment by Kludex** [comment]
It's not something we'll be able to solve. We have already done everything that we could.

You can follow the discussions and solution already merged:
- https://github.com/tiangolo/fastapi/issues/713 (November 2019)
- https://github.com/tiangolo/fastapi/issues/1691 (July 2020)
- https://github.com/tiangolo/fastapi/pull/1626 (July 2020)

To summarize:
![image](https://user-images.githubusercontent.com/7353520/92027004-30d40280-ed62-11ea-8d9f-4226bff8a55c.png)

**Comment by ycd** [comment]
The HTTP GET method requests a representation of the specified resource. Requests using GET should only retrieve data.

<table>
<tr>
<td>
Request has body
</td>
<td> NO</td>
</tr>
<tr><td>
Successful response has body
</td><td>YES</td></tr>
</table>

The [**HTTP specification**](https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.3) says in section 4.3

> A message-body MUST NOT be included in a request if the specification of the request method (section 5.1.1) does not allow sending an entity-body in requests.

**Comment by ArcLightSlavik** [comment]
@ycd The HTTP specification your using is from 1991. The updated one is RFC 7231 where it **is** allowed to include a body.
```
   A payload within a GET request message has no defined semantics;
   sending a payload body on a GET request might cause some existing
   implementations to reject the request.
```
So generally it is allowed but discouraged.
OpenAPI take on it: https://github.com/OAI/OpenAPI-Specification/pull/1937

**Comment by ycd** [comment]
An update to this issue, i found [this](https://swagger.io/docs/specification/describing-request-body/) when looking for something else, 

> GET, DELETE and HEAD are no longer allowed to have request body because it does not have defined semantics as per [RFC 7231.](https://tools.ietf.org/html/rfc7231#section-4.3.1)

since FastAPI is based on OpenAPI specification i do not think we are able to this.

**Comment by miraculixx** [comment]
> GET, DELETE and HEAD are no longer allowed to have request body because it does not have defined semantics as per RFC 7231.

That's true for OpenAPI 3.0, but seems to be changed in 3.1 as [this PR suggests](https://github.com/OAI/OpenAPI-Specification/pull/2117). It was kept in 3.0 only because it did't fit the semantics of a [patch release](https://github.com/OAI/OpenAPI-Specification/pull/2114)

> In other cases where the HTTP spec is vague (such as [GET](https://tools.ietf.org/html/rfc7231#section-4.3.1), [HEAD](https://tools.ietf.org/html/rfc7231#section-4.3.2) and [DELETE](https://tools.ietf.org/html/rfc7231#section-4.3.5)), `requestBody` is permitted but does not have well-defined semantics and SHOULD be avoided if possible.

**Comment by ycd** [comment]
@stevesuh the code piece works when you send a request with cURL

```curl
curl -X GET "http://127.0.0.1:8000/foo/bar" -d "{\"countryID\":1}"
```

So basically FastAPI **allows** request body with `GET` operations, however, Swagger UI doesn't support it.

They are also very strict about not supporting it. 

See: https://github.com/swagger-api/swagger-ui/issues/2136

If you want you can open another issue in Swagger UI's repository but FastAPI can not do anything about this topic since your issue is completely about the Swagger UI.

**Comment by stevesuh** [comment]
I consider the issue resolved then.

**Comment by tiangolo** [comment]
Thanks for the help here everyone! 👏  🙇 

Thanks for reporting back and closing the issue @stevesuh  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
